app/services/arcgis_api/token_keeper.rb

```diff
@@ -0,0 +1,234 @@
+module ArcgisApi
+  # Class for retrieve, refresh and manage caching of Arcgis API token.
+  # @attr_reader [connection] The underlying Faraday::Connection
+  # @attr_writer [prefetch_ttl] number of seconds to control timing of requesting a new token
+
+  # The token is saved in cache as a hash
+  # @example cache entry
+  # {
+  #   value: "XXXXXXXXXX",
+  #   sliding_expires_at: xxxxxxx,
+  #   expires_at: xxxxxxxxx,
+  # }
+  # expires_at: unix time that the token will expire, returned from
+  #   the Arcgis API generateToken call.
+  #
+  # sliding_expires_at: A time that the token does not actually
+  #   expire but used to control timing of requesting a new token before it expires.
+  #   It's initially set to expires_at - 3*prefetch_ttl. Only when sliding expiration enabled.
+  #
+  # If token synchronous fetching is disabled, a token will be fetched from cache directly
+  # or it will be a cache miss. Otherwise the thread will try to fetch/refresh
+  # the token on demand as needed.
+  #
+  # When sliding_expires_at<= current time <= sliding_expires_at + prefetch_ttl,
+  # the entry's sliding_expires_at time is updated to
+  # sliding_expires_at + prefetch_ttl and a new token is requested and saved to cache.
+  #
+  # When sliding_expires_at + prefetch_ttl < current time,
+  # a new token is requested and saved to cache.
+  #
+  # Optimistically, the token in cache will NOT expire
+  # when accessed by multi-threaded/distributed clients,
+  # since there is about prefetch_ttl - 2*prefetch_ttl
+  # length of time to generate a new API token.
+  #
+  class TokenKeeper
+    API_TOKEN_HOST = URI(IdentityConfig.store.arcgis_api_generate_token_url).host
+    API_TOKEN_CACHE_KEY = "arcgis_api_token:#{API_TOKEN_HOST}"
+    API_PREFETCH_TTL_SECONDS = 10
+
+    attr_accessor :connection_factory, :prefetch_ttl, :cache_key, :analytics
+
+    # @param [String] cache_key token cache key
+    # @param [ArcgisApi::ConnectionFactory] connection_factory
+    # @param [Number] prefetch_ttl number of seconds used to calculate a sliding_expires_at time
+    def initialize(cache_key, connection_factory, prefetch_ttl)
+      @cache_key = cache_key || API_TOKEN_CACHE_KEY
+      @connection_factory = connection_factory || ArcgisApi::ConnectionFactory.new
+      @prefetch_ttl = prefetch_ttl && prefetch_ttl > 0 ? prefetch_ttl : API_PREFETCH_TTL_SECONDS
```

Comment:
Let's switch to keyword arguments here, and also provide some default values?
```suggestion
    def initialize(cache_key: API_TOKEN_CACHE_KEY, connection_factory: ArcgisApi::ConnectionFactory.new, prefetch_ttl: API_PREFETCH_TTL_SECONDS)
      @cache_key = cache_key
      @connection_factory = connection_factory
      @prefetch_ttl = prefetch_ttl
```

Reply:
Done